# 1. chi square test
used for analyzing 2 categorical data

In [3]:
import numpy as np
import pandas as pd
import scipy.stats as stats
from scipy.stats import chi2_contingency

In [5]:
data = {"Female": [6,13,16,8],
        "Male": [7,16,15,11]}
education = ["without graduation", "College", "Bachelors Degree", "Masters Degree"]
df = pd.DataFrame(data, index = education)
df

,Female,Male
without graduation,6,7
College,13,16
Bachelors Degree,16,15
Masters Degree,8,11


**Is there any relationship between gender and higher eduction, so make null and alternate hypothesis**

H0: there is no relationship between gender and higher eduction
H1: there is an evident relationship between gender and higher eduction

In [18]:
test, p_value,dof, expected_value = chi2_contingency(df, correction=True)
print("p_value is:", p_value)
print("chi square test statistcs is:", test)
print("degree of freedom is:", dof)                # dof = (rows-1)(col-1)
print("expected values are:", expected_value)      # it will give expected value seperately for female and male resp[ectively

alpha= 0.05
if p_value<alpha:
    print("reject the null hypothesis")
else:
    print("failed to reject the null hypothesis")

p_value is: 0.9180008113875626
chi square test statistcs is: 0.5040497104563157
degree of freedom is: 3
expected values are: [[ 6.07608696  6.92391304]
 [13.55434783 15.44565217]
 [14.48913043 16.51086957]
 [ 8.88043478 10.11956522]]
failed to reject the null hypothesis


**Important point: we use chi2_contingency when we dont know the underlying distribution but we want to test whether two or more groups have same distribution or not**

Underlying distribution means the pattern/probability distribution that generates the data in the population (e.g., Normal, Binomial, or the proportions across categories). In Chi-Square example, you don't assume a specific distribution like Normal; you work with the observed and expected frequencies of categories.

In [11]:
from scipy.stats import chisquare
chisquare(df)

Power_divergenceResult(statistic=array([5.8372093 , 4.14285714]), pvalue=array([0.11980461, 0.2464456 ]))

**Explanation:** 

1. Female column: chisquare(df) asks: Are females equally distributed across the four education categories?: This is H0

Result:p_value = 0.1198 > 0.05 → Fail to reject H₀.

2. Male column: It asks: Are males equally distributed across the four education categories?: This is H0

Result:p_value = 0.2464 > 0.05 → Fail to reject H₀.

**Important point: we use chisquare when we have distribution in mind and we want to test whether group matches that distribution or not**

**conclusion:**
	
a. chisquare()	asks Main question: Does one categorical variable follow a particular expected distribution?

b. chi2_contingency() asks Main question: Are two categorical variables associated/independent?

chisquare(df):

Female → Are education frequencies equally distributed?

Male   → Are education frequencies equally distributed?

chi2_contingency(df): Is education level associated with gender?

This distinction is very important for interviews.

# 2. ANOVA Test
Analysis of Variance: used to compare means of 2 or more groups 

In [20]:
import numpy as np
import scipy.stats as stats

In [21]:
data1 = [50,60,80,90,100]
data2 = [30,80,80,20,170]
data3 = [70,60,30,160,60]

Null hypothesis (H0): All group means are equal : μ1=μ2=μ3

Alternative hypothesis (H1): At least one group mean is different.

In [22]:
mean1 = np.mean(data1)
mean2 = np.mean(data2)
mean3 = np.mean(data3)
mean1

np.float64(76.0)

In [23]:
mean2

np.float64(76.0)

In [24]:
mean3

np.float64(76.0)

**as all have same mean, now we check variance for that we apply ANOVA test using f-distribution table (f-test)**

In [27]:
f_test, p_value = stats.f_oneway(data1, data2, data3)
print("p_value is:", p_value)
print("f-statistcs is:", f_test)               

alpha= 0.05
if p_value<alpha:
    print("reject the null hypothesis")
else:
    print("failed to reject the null hypothesis")

p_value is: 1.0
f-statistcs is: 0.0
failed to reject the null hypothesis


**Explanation:**

failed to reject the null hypothesis Means There is no statistical evidence that the three population means are different.

**Very important: ANOVA uses VARIANCE to compare MEANS**

This can initially sound confusing:"Why are we using variance when we want to compare means?"

-> Because ANOVA asks: Are the differences between group means large compared with the natural variation within the groups?